### PR TITLE
[Parser]: Labeled block without do diagnostics

### PR DESCRIPTION
**Pull request by kitaisreal**
Improve diagnostics for labeled block without 'do'

More information in ticket https://bugs.swift.org/browse/SR-3867

### CODE REVIEW

**Comment by varungandhi-apple** [comment]
@kitaisreal, going forward, could you add some more information in the commit messages? Right now, the two commit messages in this PR are very similar. There are a couple of options, whichever works for you: (1) make changes as suggested, rebase + force-push (review is still going on) or (2) make changes as suggested in new commits, rebase + force-push after changes are approved, retest and merge.

**Comment by kitaisreal** [comment]
Hi, @varungandhi-apple lets make changes in new commits after review, then rebase, force push and retest.

**Comment by rintaro** [comment]
I was thinking of another strategy for this.
- `isStartOfStmt()` returns `true` for `identifier ':' '{'`
- In `parseStmt()`, add `tok::l_brace` case to handle `identifier ':' '{'` case.
- Add an option to `parseStmtDo()` to indicate consuming `do` should be skipped.
- In `parseStmtDo()`, if the flag is true, use the location of `{` for `do` keyword location.

Using this strategy, we can stop emitting `use of unresolved identifier` diagnostics, and parse the body as the body of the 'do' statement (instead of skipping them).
What do you think?

**Comment by kitaisreal** [comment]
Hi, @rintaro totally agree. Fixed.

File: lib/Parse/ParseStmt.cpp [hunk 1]
```diff
@@ -637,6 +641,12 @@ ParserResult<Stmt> Parser::parseStmt() {
     if (LabelInfo) diagnose(LabelInfo.Loc, diag::invalid_label_on_stmt);
     if (tryLoc.isValid()) diagnose(tryLoc, diag::try_on_stmt, Tok.getText());
     return parseStmtPoundAssert();
+  case tok::l_brace:
+    if (tryLoc.isValid()) diagnose(tryLoc, diag::try_on_stmt, Tok.getText());
+    SourceLoc colonLoc = Lexer::getTokenAtLocation(SourceMgr, PreviousLoc).getLoc();
+    diagnose(colonLoc, diag::labeled_block_needs_do)
+      .fixItInsert(colonLoc, "do");
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: This insert `do` before the `:`. I believe `fixItInsert(Tok.getLoc(), "do ")` will do the job. And I think diagnostics location should be `{` loc too. So we don't need `colonLoc` at all.

File: lib/Parse/ParseStmt.cpp [hunk 2]
```diff
@@ -93,6 +93,10 @@ bool Parser::isStartOfStmt() {
     Parser::BacktrackingScope backtrack(*this);
     consumeToken(tok::identifier);
     consumeToken(tok::colon);
+
+    if (Tok.is(tok::l_brace)) {
+      return true;
+    }
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Could you add a comment why we treat `LABEL: {` as a start of statement?

File: lib/Parse/ParseStmt.cpp [hunk 3]
```diff
@@ -1875,9 +1885,9 @@ ParserResult<Stmt> Parser::parseStmtRepeat(LabeledStmtInfo labelInfo) {
 ///   stmt-do:
 ///     (identifier ':')? 'do' stmt-brace
 ///     (identifier ':')? 'do' stmt-brace stmt-catch+
-ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo) {
+ParserResult<Stmt> Parser::parseStmtDo(LabeledStmtInfo labelInfo, bool shouldSkipDoTokenConsume) {
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns.

File: lib/Parse/ParseStmt.cpp [hunk 4]
```diff
@@ -1875,9 +1885,9 @@ ParserResult<Stmt> Parser::parseStmtRepeat(LabeledStmtInfo labelInfo) {
...
   SyntaxContext->setCreateSyntax(SyntaxKind::DoStmt);
-  SourceLoc doLoc = consumeToken(tok::kw_do);
+  SourceLoc doLoc = shouldSkipDoTokenConsume ? Tok.getLoc() : consumeToken(tok::kw_do);
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: 80 columns.

**Review by rintaro — CHANGES_REQUESTED** [review]
This looks great overall! Just a few comments.

File: lib/Parse/ParseStmt.cpp (hunk 2 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Sure. Added.

File: lib/Parse/ParseStmt.cpp (hunk 3 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 4 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 1 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Sure. Updated.

File: lib/Parse/ParseStmt.cpp [hunk 5]
```diff
@@ -637,6 +643,12 @@ ParserResult<Stmt> Parser::parseStmt() {
     if (LabelInfo) diagnose(LabelInfo.Loc, diag::invalid_label_on_stmt);
     if (tryLoc.isValid()) diagnose(tryLoc, diag::try_on_stmt, Tok.getText());
     return parseStmtPoundAssert();
+  case tok::l_brace:
+    if (tryLoc.isValid()) diagnose(tryLoc, diag::try_on_stmt, Tok.getText());
+    SourceLoc colonLoc = Tok.getLoc();
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Sorry for nitpicking:
```suggestion
    SourceLoc braceLoc = Tok.getLoc();
```

File: lib/Parse/ParseStmt.cpp [hunk 6]
```diff
@@ -637,6 +643,12 @@ ParserResult<Stmt> Parser::parseStmt() {
...
+    diagnose(colonLoc, diag::labeled_block_needs_do)
+      .fixItInsert(colonLoc, "do");
```
Comment (rintaro) [inline lib/Parse/ParseStmt.cpp]: Please add a space after `do`:
```suggestion
      .fixItInsert(colonLoc, "do ");
```

Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

File: lib/Parse/ParseStmt.cpp (hunk 5 above)
Comment (kitaisreal) [reply lib/Parse/ParseStmt.cpp]: Fixed.

**Comment by rintaro** [comment]
@swift-ci Please smoke test

**Comment by rintaro** [comment]
@swift-ci Please smoke test macOS

**Review by rintaro — APPROVED** [review]
Thank you @kitaisreal !

**Pull request merged**